### PR TITLE
Fixing scenario where a new Contact with no Address fields creates an Address

### LINKED ISSUE #6759: Default Primary Address Type with Blank Address Fields Cause Blank Address Record to be Inserted and Deleted
If a customer creates a Contact that has a value in the Primary Address Type field (npe01__Primary_Address_Type__c) but no data in any Mailing Address fields, NPSP will still create a blank Address record and then delete it, causing unnecessary SOQL queries and DML operations. Some customers have set a default picklist value on this field, so the possibility of this scenario is higher when they insert Contacts declaratively and programmatically.

### Steps to Repeat
* Go to Setup --> Object Manager --> Contact --> Fields & Relationships --> Primary Address Type. Edit the "Home" Picklist Value and set it as a default.
* Create a Contact with a First Name and Last Name. The Primary Address Type will default to "Home". Click Save. 

### Root Cause
When ADDR_Contact_TDTM runs in AfterInsert context, we call ContactAdapter.onAfterInsert(), which calls createAddressesForContacts(). We loop through each Contact in getAddressCreationQueueInAfterContext() and we check to see if isAddressChanged(),isMovingToNewHoushold() or if undeliverable status changed.  In isAddressChanged(), the ContactService.isContactAddressChanged() method returns a boolean
```
return (!isContactAddressEmpty(con1) ||
con1.npe01__Primary_Address_Type__c != null);)
```
isContactAddressEmpty() returns true since all mailing fields are blank, but the second line of the return statement is true because of the default picklist value, so we ultimately add the Contact to the address creation queue for insertion.

### Limits Comparison
In 3.209, with a default picklist value set  on Primary Address Type and no mailing address fields populated:
* 27 SOQL queries
* 5 DML Statements (3 are unnecessary - the address insert,  address delete, and update)
* 5 DML rows

In 3.209, with a blank Primary Address Type and no mailing address fields populated:
* 17 SOQL queries
* 2 DML statements
* 2 DML rows

This behavior was also present in 3.208, but order of execution was different. There was an additional DML update and 2 fewer SOQL queries. 
* 25 SOQL queries
* 6 DML statements
* 6 DML rows

### Workaround
If you have a default picklist value configured on the Primary Address Type field and are experiencing any npsp SOQL query or DML limits being hit in unit tests or in other areas where you may insert Contacts with no mailing address fields populated, clear the value in the npe01__Primary_Address_Type__c field. (In the UI, set the value to --None--; programmatically, set the value to '').

### PR DESCRIPTION
**Pull request by lparrott**
# Critical Changes

# Changes

- We fixed a scenario where inserting or updating a Contact with a Primary Address Type, but no Address fields, would unnecessarily create, then delete, an Address record.

# Issues Closed

Fixes #6759 

# Community Ideas Delivered

# Features Intended for Future Release

# Features for Elevate Customers

# New Metadata

# Deleted Metadata

### CODE REVIEW

File: force-app/main/adapter/in/sobjects/contact/ContactAdapter.cls [hunk 1]
```diff
@@ -477,7 +477,7 @@ public inherited sharing class ContactAdapter extends fflib_SObjects {
         for (Integer i = 0; i < getRecords().size(); i++) {
             Contact newRecord = (Contact) getRecords()[i];
 
-            if (contactService.isContactAddressChanged(newRecord, oldVersionOf(newRecord))) {
+            if (new NPSP_Contact(newRecord, oldVersionOf(newRecord)).isAddressChanged()) {
```
Comment (lparrott) [inline force-app/main/adapter/in/sobjects/contact/ContactAdapter.cls:480]: I'm wondering if there's a more efficient (or more clear) way to do this

File: force-app/tdtm/classes/TDTM_DMLgt200_TEST.cls [hunk 2]
```diff
@@ -363,15 +363,15 @@ private class TDTM_DMLgt200_TEST {
...
-        Test.startTest();
-
         // Create more than 200 Contact records and then validate that all of the Household Accounts were created
         List<Contact> contacts = createMultipleTestContacts(insertCnt);
         insertWithAllowDupes(contacts);
 
         System.assertEquals(insertCnt, [SELECT Count() FROM Account WHERE npe01__SYSTEM_AccountType__c = :CAO_Constants.HH_ACCOUNT_TYPE],
                 'There should be exactly ' + insertCnt.format() + ' Accounts');
 
+        Test.startTest();
```
Comment (lparrott) [inline force-app/tdtm/classes/TDTM_DMLgt200_TEST.cls:373]: This change was made to fix a CPU Limit error

File: force-app/main/service/ContactService.cls [hunk 3]
```diff
@@ -68,54 +68,6 @@ public with sharing class ContactService {
...
-                !addressServiceInstance.equalsCaseSensitive(con1.MailingStreet, con2.MailingStreet) ||
-                        !addressServiceInstance.equalsCaseSensitive(con1.MailingCity, con2
-                                .MailingCity) ||
-                        !addressServiceInstance.equalsCaseSensitive(con1.MailingState, con2
-                                .MailingState) ||
-                        !addressServiceInstance.equalsCaseSensitive(con1.MailingPostalCode, con2
-                                .MailingPostalCode) ||
-                        !addressServiceInstance.equalsCaseSensitive(con1.MailingCountry, con2
-                                .MailingCountry) ||
-                        !addressServiceInstance.equalsCaseSensitive(con1
```
Comment (lparrott) [inline force-app/main/service/ContactService.cls:103]: Just to note, the previous logic considered an update to Primary Address Type a "change", even if nothing else was changed or filled in, which is what our new logic also does.

Comment (lparrott) [reply force-app/main/service/ContactService.cls:103]: Upon further testing, I decided to check for a non-empty address to allow Primary Address Type to cause a change, it was adding unneeded DML just like the insert case that we're trying to resolve!

**Review by andrewyu-salesforce — APPROVED** [review]
The change looks great. Unnecessary address soql and dml count are all removed.

**Review by dfuller114 — APPROVED** [review]
looks good!

**Comment by andrewyu-salesforce** [comment]
[WI-10252900](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000UTBZvYAP/view)

**Pull request merged**